src/qtism/runtime/tests/AssessmentTestSession.php

```diff
@@ -2406,10 +2406,31 @@ protected function nextRouteItem($ignoreBranchings = false, $ignorePreConditions
         $stop = false;
 
         while ($route->valid() === true && $stop === false) {
+            $currentRouteItem = $route->current();
+            $branchRules = $currentRouteItem->getBranchRules();
+
+            if ($branchRules->count() === 0) {
+                $currentSection = $currentRouteItem->getAssessmentSection();
+                $sectionItems = $route->getRouteItemsByAssessmentSection($currentSection)->getArrayCopy();
+
+                if (end($sectionItems) === $currentRouteItem) {
+                    $branchRules = $currentSection->getBranchRules();
+                }
+
+                if ($branchRules->count() === 0) {
+                    $testPartItems = $route->getCurrentTestPartRouteItems()->getArrayCopy();
+
+                    if (end($testPartItems) === $currentRouteItem) {
+                        $branchRules = $currentRouteItem->getTestPart()->getBranchRules();
+                    }
+                }
+            }
+
+            $numberOfBranchRules = $branchRules->count();
```

Comment:
Maybe we can move it as a function to `RouteItem::getEffectiveBranchingRules`? 

```
private function getEffectiveBranchingRules($routeItem): ?Collection {
    $currentRouteItem = $route->current();
    $branchRules = $currentRouteItem->getBranchRules();
    
    if ($branchRules->count() > 0) {
        return $branchRules;
    }
    
    $currentSection = $currentRouteItem->getAssessmentSection();
    $sectionItems = $route->getRouteItemsByAssessmentSection($currentSection)->getArrayCopy();
    
    if (end($sectionItems) === $currentRouteItem) {
            $branchRules = $currentSection->getBranchRules();
    }
    
    if ($branchRules->count() > 0) {
        return $branchRules;
    }
    
    $testPartItems = $route->getCurrentTestPartRouteItems()->getArrayCopy();
        
    if (end($testPartItems) === $currentRouteItem) {
        return $branchRules = $currentRouteItem->getTestPart()->getBranchRules();
    }
}
```

Reply:
Thanks! Done: https://github.com/oat-sa/qti-sdk/pull/370/commits/c623ab4dbbb3df9acfefcbe23967c5c444c8b7f0